src/storage/src/render/persist_sink.rs

```diff
@@ -932,439 +935,574 @@ where
+                                .or_default();
+
+                            batches.finished.push(FinishedBatch {
+                                batch: write.batch_from_transmittable_batch(batch.batch),
+                                data_ts: batch.data_ts,
+                            });
+                            batches.batch_metrics += &batch.metrics;
                         }
+                        continue 'consumer;
+                    }
+                    Event::Progress(frontier) => {
+                        batches_frontier = frontier;
                     }
                 }
-                else => {
-                    // All inputs are exhausted, so we can shut down.
-                    return Ok(());
-                }
-            };
-
-            // Peel off any batches that are not beyond the frontier
-            // anymore.
-            //
-            // It is correct to consider batches that are not beyond the
-            // `batches_frontier` because it is held back by the writer
-            // operator as long as a) the `batch_description_frontier` did
-            // not advance and b) as long as the `desired_frontier` has not
-            // advanced to the `upper` of a given batch description.
-
-            let mut done_batches = in_flight_descriptions
-                .iter()
-                .filter(|(lower, _upper)| !PartialOrder::less_equal(&batches_frontier, lower))
-                .cloned()
-                .collect::<Vec<_>>();
-
-            trace!(
-                "persist_sink {collection_id}/{shard_id}: \
-                    append_batches: in_flight: {:?}, \
-                    done: {:?}, \
-                    batch_frontier: {:?}, \
-                    batch_description_frontier: {:?}",
-                in_flight_descriptions,
-                done_batches,
-                batches_frontier,
-                batch_description_frontier
-            );
-
-            // Append batches in order, to ensure that their `lower` and
-            // `upper` line up.
-            done_batches.sort_by(|a, b| {
+            }
+            else => {
+                // All inputs are exhausted, so we can shut down.
+                return Ok(());
+            }
+        };
+        // Peel off any batches that are not beyond the frontier
+        // anymore.
+        //
+        // It is correct to consider batches that are not beyond the
+        // `batches_frontier` because it is held back by the writer
+        // operator as long as a) the `batch_description_frontier` did
+        // not advance and b) as long as the `desired_frontier` has not
+        // advanced to the `upper` of a given batch description.
+
+        let mut done_batches = in_flight_descriptions
+            .iter()
+            .filter(|(lower, _upper)| !PartialOrder::less_equal(&batches_frontier, lower))
+            .cloned()
+            .sorted_by(|a, b| {
```

Comment:
This is `Itertools::sorted_by` which will create a temporary vector with all the elements of the iterator, sort them, and then pass the resulting vector as an iterator to `collect`, which will allocate a brand new vector to put the elements in.

I think what you want to do is collect first and then sort the resulting vector in a separate statement